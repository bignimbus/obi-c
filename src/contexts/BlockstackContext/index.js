import React, {
  useState,
  useEffect,
  createContext,
} from 'react';
import PropTypes from 'prop-types';
import {
  Person,
  signUserOut,
  loadUserData,
  isUserSignedIn,
  isSignInPending,
  redirectToSignIn,
  handlePendingSignIn,
} from 'blockstack';

const BlockstackContext = createContext();

export const PENDING = 'pending';
export const LOADING = 'loading';
export const AUTHENTICATED = 'authenticated';

export const BlockstackContextProvider = ({ children }) => {
  const [user, setUser] = useState(null);
  const [authState, setAuthState] = useState(PENDING);

  const signIn = () => {
    redirectToSignIn();
  };

  const signOut = () => {
    signUserOut(window.location.origin);
  }

  const verifyAuthStatus = async () => {
    if (isUserSignedIn()) {
      const { profile } = loadUserData();
      setUser(new Person(profile));
      setAuthState(AUTHENTICATED);
    } else if (isSignInPending()) {
      try {
        setAuthState(LOADING);
        const { profile } = await handlePendingSignIn();
        setUser(new Person(profile));
        setAuthState(AUTHENTICATED);
      } catch (e) {
        console.warn(e);
      }
    }
  };

  useEffect(() => {
    if (user) return;
    verifyAuthStatus();
  });

  return (
    <BlockstackContext.Provider
      value={{
        user,
        signIn,
        signOut,
        authState,
      }}
    >
      { children }
    </BlockstackContext.Provider>
  );
};

BlockstackContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default BlockstackContext;
