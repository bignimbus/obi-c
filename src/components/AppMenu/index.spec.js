import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import BlockstackContext from '../../contexts/BlockstackContext';
import { default as AppMenu } from '.';

const renderComponent = contextValue => renderer.create(
  <BlockstackContext.Provider value={contextValue}>
    <MemoryRouter>
      <AppMenu />
    </MemoryRouter>
  </BlockstackContext.Provider>,
);

describe('AppMenu', () => {
  describe('#render', () => {
    test('unauthenticated user', () => {
      const element = renderComponent({
        user: null,
        signIn: jest.fn(),
        authState: 'pending',
      });
      expect(element.toJSON()).toMatchSnapshot();
    });

    test('authenticated user fetching data', () => {
      const element = renderComponent({
        user: null,
        signIn: jest.fn(),
        authState: 'loading',
      });
      expect(element.toJSON()).toMatchSnapshot();
    });

    test('authenticated user', () => {
      const element = renderComponent({
        user: {
          givenName: jest.fn(() => 'Jeff'),
        },
        signIn: jest.fn(),
        authState: 'authenticated',
      });
      expect(element.toJSON()).toMatchSnapshot();
    });
  });
});
