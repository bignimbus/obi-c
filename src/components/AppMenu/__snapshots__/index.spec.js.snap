// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppMenu #render authenticated user 1`] = `
<div
  className="app-menu"
>
  <nav
    className="app-menu__nav"
  >
    <header
      className="app-menu__header"
    >
      <div>
        Jeff
      </div>
    </header>
    <section
      className="app-menu__section"
    >
      <ul
        className="app-menu__ul"
      >
        <li
          className="app-menu__li"
        >
          <a
            href="/remind"
            onClick={[Function]}
          >
            Remind me
          </a>
        </li>
        <li
          className="app-menu__li"
        >
          <a
            href="/feed"
            onClick={[Function]}
          >
            Feed obi-c
          </a>
        </li>
        <li
          className="app-menu__li"
        >
          <a
            href="/play"
            onClick={[Function]}
          >
            Play with obi-c
          </a>
        </li>
      </ul>
    </section>
  </nav>
</div>
`;

exports[`AppMenu #render authenticated user fetching data 1`] = `
<div
  className="app-menu"
>
  <nav
    className="app-menu__nav"
  >
    <header
      className="app-menu__header"
    >
      <div>
        <section>
          <button
            disabled={true}
            onClick={[MockFunction]}
          >
            <span>
              Connecting to Blockstack...
            </span>
          </button>
        </section>
        <footer>
          <button>
            Learn more
          </button>
        </footer>
      </div>
    </header>
    <section
      className="app-menu__section"
    >
      <ul
        className="app-menu__ul"
      >
        <li
          className="app-menu__li"
        >
          <a
            href="/remind"
            onClick={[Function]}
          >
            Remind me
          </a>
        </li>
        <li
          className="app-menu__li"
        >
          <a
            href="/feed"
            onClick={[Function]}
          >
            Feed obi-c
          </a>
        </li>
        <li
          className="app-menu__li"
        >
          <a
            href="/play"
            onClick={[Function]}
          >
            Play with obi-c
          </a>
        </li>
      </ul>
    </section>
  </nav>
</div>
`;

exports[`AppMenu #render unauthenticated user 1`] = `
<div
  className="app-menu"
>
  <nav
    className="app-menu__nav"
  >
    <header
      className="app-menu__header"
    >
      <div>
        <section>
          <button
            disabled={false}
            onClick={[MockFunction]}
          >
            <span>
              Connect to Blockstack
            </span>
          </button>
        </section>
        <footer>
          <button>
            Learn more
          </button>
        </footer>
      </div>
    </header>
    <section
      className="app-menu__section"
    >
      <ul
        className="app-menu__ul"
      >
        <li
          className="app-menu__li"
        >
          <a
            href="/remind"
            onClick={[Function]}
          >
            Remind me
          </a>
        </li>
        <li
          className="app-menu__li"
        >
          <a
            href="/feed"
            onClick={[Function]}
          >
            Feed obi-c
          </a>
        </li>
        <li
          className="app-menu__li"
        >
          <a
            href="/play"
            onClick={[Function]}
          >
            Play with obi-c
          </a>
        </li>
      </ul>
    </section>
  </nav>
</div>
`;
